src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -2053,24 +2057,69 @@ UpdateShardPlacement(PlacementUpdateEvent *placementUpdateEvent,
  * ExecuteCriticalCommandInSeparateTransaction runs a command in a separate
  * transaction that is commited right away. This is useful for things that you
  * don't want to rollback when the current transaction is rolled back.
+ * Set true to 'useExclusiveTransactionBlock' to initiate a BEGIN and COMMIT statements.
  */
 void
-ExecuteRebalancerCommandInSeparateTransaction(char *command)
+ExecuteRebalancerCommandInSeparateTransaction(char *command, bool
+											  useExclusiveTransactionBlock)
 {
 	int connectionFlag = FORCE_NEW_CONNECTION;
 	MultiConnection *connection = GetNodeConnection(connectionFlag, LocalHostName,
 													PostPortNumber);
 	StringInfo setApplicationName = makeStringInfo();
-	appendStringInfo(setApplicationName, "SET application_name TO %s",
+	appendStringInfo(setApplicationName, "SET LOCAL application_name TO %s;",
 					 CITUS_REBALANCER_NAME);
 
+	if (PropagateSessionSettingsForLoopbackConnection)
+	{
+		StringInfo setStatements = GetSetStatementsForNewConnections();
+		appendStringInfoString(setApplicationName, setStatements->data);
+	}
+
+	if (useExclusiveTransactionBlock)
+	{
+		ExecuteCriticalRemoteCommand(connection, "BEGIN;");
+	}
+
 	ExecuteCriticalRemoteCommand(connection, setApplicationName->data);
 	ExecuteCriticalRemoteCommand(connection, command);
 
+	if (useExclusiveTransactionBlock)
+	{
+		ExecuteCriticalRemoteCommand(connection, "COMMIT;");
+	}
+
 	CloseConnection(connection);
 }
 
 
+/*
+ * GetSetStatementsForNewConnections returns a string of SET statements to
+ * be executed in new connections to worker nodes.
+ */
+static StringInfo
+GetSetStatementsForNewConnections(void)
+{
```

Comment:
What about if I do:

```SQL
BEGIN;
SET LOCAL citus.max_adaptive_executor_pool_size TO 5;
SELECT rebalance_table_shards();
```

Or, `SET SESSION ...`

Can we catch these with `PGC_S_SESSION` or should we add few more sources as well?

Reply:
Yes, we can catch those, as well.

```sql 
gokhan=# BEGIN;
BEGIN
gokhan=*# SET citus.log_remote_commands to ON;
SET
gokhan=*# SET LOCAL citus.max_adaptive_executor_pool_size TO 5;
SET
gokhan=*# SET SESSION citus.shard_count to 16;
SET
gokhan=*# select citus_drain_node('localhost', 9701, shard_transfer_mode := 'force_logical');
NOTICE:  issuing BEGIN;
DETAIL:  on server gokhan@localhost:9700 connectionId: 24
NOTICE:  issuing SET LOCAL application_name TO citus_rebalancer;SET LOCAL citus.log_remote_commands TO 'on';SET LOCAL citus.max_adaptive_executor_pool_size TO '5';SET LOCAL citus.shard_count TO '16';
DETAIL:  on server gokhan@localhost:9700 connectionId: 24`